lib/rust/ensogl/core/src/debug/monitor.rs

```diff
@@ -931,3 +884,113 @@ stats_sampler!(
     0,
     1.0
 );
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    use assert_approx_eq::assert_approx_eq;
+
+
+    #[test]
+    fn frame_time() {
+        // Note: 60 FPS means there's 16.6(6) ms budget for 1 frame. The test will be written under
+        // assumption we're trying to be around this FPS.
+
+        let stats: Stats = default();
+        let sampler: Box<dyn Sampler> = Box::new(FrameTime::new(&stats));
+        let mut t = 0.0f64;
+
+        // Frame 1: simulate we managed to complete the work in 10ms, and then we wait 6ms before
+        // starting next frame.
+        stats.begin_frame(t);
+        t += 10.0;
+        stats.end_frame(t);
+        assert_approx_eq!(sampler.value(), 10.0);
+        assert!(matches!(sampler.check(), ValueCheck::Correct));
+        t += 6.0;
+
+        // Frame 2: simulate we managed to complete the work in 5ms, and then we wait 11ms before
+        // starting next frame.
+        stats.begin_frame(t);
+        t += 5.0;
+        stats.end_frame(t);
+        assert_approx_eq!(sampler.value(), 5.0);
+        assert!(matches!(sampler.check(), ValueCheck::Correct));
+        t += 11.0;
+
+        // Frame 3: simulate we went over the budget of 16.6(6) ms, at 30ms. No extra delay
+        // afterwards before starting next frame.
+        stats.begin_frame(t);
+        t += 30.0;
+        stats.end_frame(t);
+        assert_approx_eq!(sampler.value(), 30.0);
+        assert!(matches!(sampler.check(), ValueCheck::Warning));
+
+        // Frame 4: simulate a really slow frame (1000ms), crossing the configured error threshold.
+        stats.begin_frame(t);
+        t += 1000.0;
+        stats.end_frame(t);
+        assert_approx_eq!(sampler.value(), 1000.0);
+        assert!(matches!(sampler.check(), ValueCheck::Error));
+    }
+
+    #[test]
+    fn fps() {
+        // Note: 60 FPS means there's 16.6(6) ms budget for 1 frame. The test will be written under
+        // assumption we're trying to be around this FPS.
+
+        let stats: Stats = default();
+        let sampler: Box<dyn Sampler> = Box::new(Fps::new(&stats));
+        // Note: we can't use t=0.0 here, as this would throw off inner logic of Fps.
+        let mut t = 123.0f64;
+
+        // Frame 1: simulate we managed to complete the work in 10ms, and then we wait 6ms before
+        // starting next frame.
+        stats.begin_frame(t);
+        t += 10.0;
+        stats.end_frame(t);
+        // FPS takes into account delays between frames, so it is only available for *previous*
+        // frame. As such, after 1st frame, there was no previous frame yet, so the calculated
+        // value is expected to default to 0 ("zero FPS" is a reasonable answer at this point).
+        assert_approx_eq!(sampler.value(), 0.0);
```

Comment:
This needs to better explained. The sentence "FPS takes into account delays between frames" does not explain why this happens. Provide here a more detailed explanation, maybe with an example, please. A good example is that if for example the frame time was 1ms, we don't want to show crazy FPS, because the next frame will not be run immediately.

Reply:
I have here another, pretty serious issue – all of the samplers show the stats from the current frame, while the FPS stat shows the stats from the previous frame. This will be extremely confusing when debugging performance and this needs to be fixed. I would propose to make here a 2 step process – the monitor gets event (like function call) at the end of the frame, and gets a "draw" event on the beginning of the next frame. During the draw event, you have all information available, and all the stats can be updated in a synchronous manner.

Reply:
@wdanilo Based on my understanding of the idea above, I would propose a slightly modified solution, where `StatsData` becomes available only as a result of `Stats::begin_frame(&self, time: f64) -> Option<StatsData>` (which would be returning previous frame's `StatsData` snapshot), evaluating to `None` on first call (because 1st frame does not have a predecessor). A `for` loop sampling and postprocessing stats data, and the `Monitor::draw()` call would then be moved from `Monitor::end()` to `Monitor::begin()`, such that the drawing would be done based on data loaded from the `StatsData` object. This would also imply that `Sampler` could and should lose a reference to `Stats`, instead being passed a `StatsData` object for sampling & postprocessing, and thus possibly become stateless. As a result, the Monitor Panel would be drawn at the beginning of a frame (instead of at the end) - following your proposal above - and would draw synchronized stats of the previous frame, except for 1st frame when it would need to skip the drawing operation, having received `None` from `begin_frame`.

The above proposal would also resolve a few other concerns expressed in this review (resolving the issue with "0.0" time being treated specially, as well as removing a need to add an extra conditional value to the `stats_sampler!` macro).

Given that the above proposal would necessitate refactoring in: `Monitor`, `Sampler` trait + `stats_sampler!` macro, `PanelData`, `Stats` and `StatsData`, I estimate the time this change would take at 1-2 days, taking into account also initial Q/A done as part of implementation, and possibly new unit tests. Based on that, I would like to ask for a recommendation how to proceed:

 - a) It can be noticed, that the issue we're discussing here is not a regression introduced by this PR, but it was present in the code before this PR. As such, it is a bug we need to solve regardless if we merge this PR or not. Taking this into account, we could split off a separate Bug task for that in Pivotal, in the Profiling EPIC, and remove those concerns from the current PR review, focusing in it just on refactoring current code, and prioritizing the Bug separately in the EPIC.
 - b) We could keep it as part of this PR, taking into consideration that together with remaining changes requested in this PR review, it will delay delivery of this PR and Task by est. 2 days.

What is your suggestion how to proceed with this? I consulted with @MichaelMauderer, who leans toward adding a new task, but also suggested I ask for your opinion?

Reply:
I feel that the solution a) would just make it way longer and way more painful than necessary. We are refactoring a code, we see that it does not work and it was pointed in the review. Merging it now, describing a new issue, starting working on it ... makes the whole process super formal, and would force us to spend so much additional time just on making "paperwork" that does not bring us to the final solution.

My question here is – what is the delivery time difference between points a) and b)? How much additional time we will waste when choosing point a)?

Regarding the proposed changes, I was not checking them deeply, but they seem reasonable on the first glance. The only thing I'd change is that I'd not return `Option<StatsData>`, just `StatsData` - for the first time it can be a default value. It makes sense - the first frame has 0 fps, 0 men usage, etc.

Reply:
**Re. 1:** I attempted to improve the comment; would you find the new phrasing better?

**Re. 2:** As we discussed further, I created https://www.pivotaltracker.com/story/show/181140499 and plan to fix the calculation bug in `Fps` code (it was not throwing/panicking, but it was skipping some calculations - I now understand that the original phrasing I used in the comment here could be understood in a different way than I intended), which existed in old code in the same shape, through work on that Bug.